File: src/lib/ansi.ts
/**
 * Shared ANSI color utilities.
 * Zero dependencies — uses raw escape codes.
 * Respects NO_COLOR (https://no-color.org) and non-TTY environments.
 */

type ColorFn = (s: string) => string;

const noColor = Boolean(process.env.NO_COLOR) || (!process.stdout.isTTY && !process.stderr.isTTY);

function wrap(code: string): ColorFn {
  if (noColor) return (s: string) => s;
  return (s: string) => `\x1b[${code}m${s}\x1b[0m`;
}

/** Apply bold formatting. */
export const bold: ColorFn = wrap('1');
/** Apply dim formatting. */
export const dim: ColorFn = wrap('2');

/** Apply red foreground color. */
export const red: ColorFn = wrap('31');
/** Apply green foreground color. */
export const green: ColorFn = wrap('32');
/** Apply yellow foreground color. */
export const yellow: ColorFn = wrap('33');
/** Apply cyan foreground color. */
export const cyan: ColorFn = wrap('36');
/** Apply white foreground color. */
export const white: ColorFn = wrap('37');
/** Apply magenta foreground color. */
export const magenta: ColorFn = wrap('35');
/** Apply gray foreground color. */
export const gray: ColorFn = wrap('90');

/** Apply cyan background color. */
export const bgCyan: ColorFn = wrap('46');

/**
 * Strip all ANSI escape sequences from a string.
 */
export function stripAnsi(str: string): string {
  // eslint-disable-next-line no-control-regex
  return str.replace(/\x1b\[[0-9;]*m/g, '');
}

File: crates/ecc-domain/src/ansi.rs
//! Zero-dependency ANSI color utilities.
//! Respects `NO_COLOR` env var convention (checked at call site).

use std::sync::LazyLock;

static RE_ANSI: LazyLock<regex::Regex> =
    LazyLock::new(|| regex::Regex::new(r"\x1b\[[0-9;]*m").expect("valid regex"));

/// Controls whether ANSI color codes are emitted.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorMode {
    /// Emit ANSI escape codes.
    Enabled,
    /// Emit plain text with no escape codes.
    Disabled,
}

impl From<bool> for ColorMode {
    fn from(enabled: bool) -> Self {
        if enabled {
            ColorMode::Enabled
        } else {
            ColorMode::Disabled
        }
    }
}

/// Wrap text in bold ANSI escape codes.
pub fn bold(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "1", mode.into())
}

/// Wrap text in dim ANSI escape codes.
pub fn dim(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "2", mode.into())
}

/// Wrap text in red ANSI escape codes.
pub fn red(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "31", mode.into())
}

/// Wrap text in green ANSI escape codes.
pub fn green(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "32", mode.into())
}

/// Wrap text in yellow ANSI escape codes.
pub fn yellow(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "33", mode.into())
}

/// Wrap text in cyan ANSI escape codes.
pub fn cyan(text: &str, mode: impl Into<ColorMode>) -> String {
    wrap(text, "36", mode.into())
}

fn wrap(text: &str, code: &str, mode: ColorMode) -> String {
    if mode == ColorMode::Enabled {
        format!("\x1b[{code}m{text}\x1b[0m")
    } else {
        text.to_string()
    }
}

/// Strip all ANSI escape sequences from a string.
pub fn strip_ansi(text: &str) -> String {
    RE_ANSI.replace_all(text, "").into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bold_enabled_wraps_text() {
        assert_eq!(bold("hello", true), "\x1b[1mhello\x1b[0m");
    }

    #[test]
    fn bold_disabled_returns_plain() {
        assert_eq!(bold("hello", false), "hello");
    }

    #[test]
    fn red_enabled_wraps_text() {
        assert_eq!(red("err", true), "\x1b[31merr\x1b[0m");
    }

    #[test]
    fn strip_ansi_removes_codes() {
        let colored = bold(&red("hello", true), true);
        assert_eq!(strip_ansi(&colored), "hello");
    }

    #[test]
    fn strip_ansi_noop_on_plain() {
        assert_eq!(strip_ansi("plain text"), "plain text");
    }

    #[test]
    fn all_colors_produce_correct_codes() {
        assert!(green("x", true).contains("[32m"));
        assert!(yellow("x", true).contains("[33m"));
        assert!(cyan("x", true).contains("[36m"));
        assert!(dim("x", true).contains("[2m"));
    }

    #[test]
    fn color_mode_enum() {
        // ColorMode::Enabled produces ANSI escape codes
        assert_eq!(bold("x", ColorMode::Enabled), "\x1b[1mx\x1b[0m");
        // ColorMode::Disabled produces plain text
        assert_eq!(bold("x", ColorMode::Disabled), "x");
        // From<bool>: true maps to Enabled, false maps to Disabled
        assert_eq!(ColorMode::from(true), ColorMode::Enabled);
        assert_eq!(ColorMode::from(false), ColorMode::Disabled);
    }
}
